python/cudf/cudf/_lib/join.pyx

```diff
@@ -1,222 +1,90 @@
-                rhs_view,
-                left_on_ind,
-                right_on_ind,
-                all_left_inds
-            ))
-
-    all_cols_py = columns_from_ptr(move(c_result))
-    if left_index or right_index:
-        ind_cols = OrderedDict()
-        for name, pos in zip(
-            result_index_names[::-1], result_idx_positions[::-1]
-        ):
-            ind_cols[name] = all_cols_py.pop(pos)
-        index = OrderedDict()
-        for k, v in reversed(ind_cols.items()):
-            index[k] = v
-        index = Table(index)
+# The functions below return the *gathermaps* that represent
+# the join result when joining on the keys `lhs` and `rhs`.
+
+cpdef join(Table lhs, Table rhs, how=None):
+
+    cdef pair[cpp_join.gather_map_type, cpp_join.gather_map_type] c_result
+    cdef table_view c_lhs = lhs.view()
+    cdef table_view c_rhs = rhs.view()
+
+    if how == "inner":
+        c_result = move(cpp_join.inner_join(
+            c_lhs,
+            c_rhs
+        ))
+    elif how == "left":
+        c_result = move(cpp_join.left_join(
+            c_lhs,
+            c_rhs
+        ))
+    elif how == "outer":
+        c_result = move(cpp_join.full_join(
+            c_lhs,
+            c_rhs
+        ))
     else:
-        index = None
-    data_ordered_dict = OrderedDict(zip(result_col_names, all_cols_py))
-    return Table(data=data_ordered_dict, index=index)
-
-
-def compute_result_col_names(lhs, rhs, how):
-    """
-    Determine the names of the data columns in the result of
-    a libcudf join, based on the original left and right frames
-    as well as the type of join that was performed.
-    """
-    if how in {"left", "inner", "outer", "leftsemi", "leftanti"}:
-        a = lhs._data.keys()
-        if how not in {"leftsemi", "leftanti"}:
-            return list(chain(a, (k for k in rhs._data.keys()
-                        if k not in lhs._data.keys())))
-        return list(a)
+        raise ValueError(f"Invalid join type {how}")
+
+    cdef Column left_rows = _gather_map_as_column(move(c_result.first))
+    cdef Column right_rows = _gather_map_as_column(move(c_result.second))
+    return left_rows, right_rows
+
+
+cpdef semi_join(Table lhs, Table rhs, how=None):
+
+    # left-semi and left-anti joins
+    cdef cpp_join.gather_map_type c_result
+    cdef table_view c_lhs = lhs.view()
+    cdef table_view c_rhs = rhs.view()
+
+    if how == "leftsemi":
+        c_result = move(cpp_join.left_semi_join(
+            c_lhs,
+            c_rhs
+        ))
+    elif how == "leftanti":
+        c_result = move(cpp_join.left_anti_join(
+            c_lhs,
+            c_rhs
+        ))
     else:
-        raise NotImplementedError(
-            f"{how} merge not supported yet"
-        )
+        raise ValueError(f"Invalid join type {how}")
+
+    cdef Column left_rows = _gather_map_as_column(move(c_result))
+    return (
+        left_rows,
+        cudf.core.column.as_column([], dtype="int32")
```

Comment:
This is to make it so that we gather no rows from the rhs, right? I think some GPU work can be avoided with a `len(gather_map)==0` check in `_gather` and just returning `[]` here. Even if not, we have the `as_column` call in gather already I think.

Reply:
One advantage of returning a `Column` here is that it leads to a consistent API. The drawback currently is that we'll call `as_column()` again on the result in `_gather`, but that _should_ be a no-op.

Reply:
I see your point - it's nice that we always return two columns, the gather map for the left and the gather map for the right. Does gather short circuit for a gather map with length zero? If not, we might consider making it so that it does. 

This is a nit admittedly. But it seems like allocating a new column and then performing a gather is a lot of stuff to do for a situation where the answer is known.

Reply:
That's a good point -- let me look into if we can avoid doing a gather altogether

Reply:
OK, I added a bit of logic in https://github.com/rapidsai/cudf/pull/7454/commits/328dafdb548f6dfdf47df4b60d67652a4b8c971d, where we now return `None` as the RHS gathermap in `semi_join`.